File: house-robber-iii/house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        
        def dfs(node):
            if not node: return 0, 0, 0
            l, ll, lr = dfs(node.left)
            r, rl, rr = dfs(node.right)
            return max(node.val + ll + lr + rl + rr, l + r), l, r
        return dfs(root)[0]
            

File: house-robber-iii/house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        if (!root) return 0;
        int l = 0, r = 0;
        return helper(root, l, r);
    }
private:
    int helper(TreeNode* node, int& l, int& r){
        if (!node) return 0;
        int ll = 0;
        int lr = 0;
        int rl = 0;
        int rr = 0;
        
        l = helper(node->left, ll, lr);
        r = helper(node->right, rl, rr);
        
        return max(node->val + ll + lr + rl + rr, l + r);
    }
};
